### PR TITLE
ADEN-9559 Add Templates 2.0

### PR DESCRIPTION
**Pull request by Bielik20**

### CODE REVIEW

File: examples/templates/template-registry-demo/script.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	'uap-2-copy',
+	{
+		first: [FirstMockHandler],
+		second: [SecondMockHandler],
+	},
+	'first',
+);
+
+templateRegistry.init('uap-2', { name: 'slot name' } as any, { slotName: 'I do not know' });
+// templateRegistry.init('uap-2-copy', { name: 'another slot' } as any, {
```
Comment (jacekbj) [inline examples/templates/template-registry-demo/script.ts]: What do you want to keep this for?

File: package.json [hunk 2]
```diff
@@ -21,6 +21,7 @@
 		"lodash": "^4.17.4",
 		"query-string": "^6.8.0",
 		"rxjs": "^6.5.3",
+		"state-charts": "1.2.1",
```
Comment (jacekbj) [inline package.json]: We will not need that anymore.

File: spec/ad-engine/services/templates-registry/template-machine.spec.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import { TemplateMachine } from '@wikia/ad-engine/services/templates-registry/template-machine';
+import { assert, expect } from 'chai';
+import { createSandbox } from 'sinon';
+import { createTemplateStaterStub, TemplateStateStub } from './template-state.stub';
```
Comment (jacekbj) [inline spec/ad-engine/services/templates-registry/template-machine.spec.ts]: Type `s/createTemplateStaterStub/createTemplateStateStub/g`.

File: spec/ad-engine/services/templates-registry/template-machine.spec.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+			new Map([['a', stateA], ['b', stateB]]) as any,
+			'a',
+		);
+	});
+
+	afterEach(() => {
+		sandbox.restore();
+	});
+
+	it('should enter initial state state after init', () => {
```
Comment (jacekbj) [inline spec/ad-engine/services/templates-registry/template-machine.spec.ts]:
```suggestion
	it('should enter initial state after init', () => {
```

File: examples/templates/template-registry-demo/script.ts (hunk 1 above)
Comment (Bielik20) [reply examples/templates/template-registry-demo/script.ts]: TBH it was for testing, to make sure those two are considered different and behave in the correct manner. There are unit tests now for that, we could remove it.

File: package.json (hunk 2 above)
Comment (Bielik20) [reply package.json]: True

File: examples/templates/template-registry-demo/first-mock-handler.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Context, TemplateStateHandler, TemplateTransition } from '@wikia/ad-engine';
+import { Injectable } from '@wikia/dependency-injection';
+
+@Injectable()
+export class FirstMockHandler implements TemplateStateHandler {
+	private id = Math.random() * 10;
```
Comment (jacekbj) [inline examples/templates/template-registry-demo/first-mock-handler.ts]:
```suggestion
	private id = Math.floor(Math.random() * 1000);
```

**Pull request merged**